### DISCUSSION TITLE
Some bugs and suggestions

### DISCUSSION THREAD

**Discussion by rllbe** [ReagentX/imessage-exporter#151] [Ideas]
1. Audio messages in html output appear as downloadable attachments instead of playable audio elements
https://github.com/ReagentX/imessage-exporter/blob/c73bc4d666da3a543708b553c9106acf89454e81/imessage-database/src/tables/attachment.rs#L100-L119
As of iOS 15.4, an audio message doesn't have a "mime_type" value("mime_type" column is empty), but it does have a "uti" value which is "com.apple.coreaudio-format".
Suggestion: use "uti" column to determine whether an attachment is really an attachment or an audio message, like this:
 ``` 
            None => {
                match &self.uti {
                    Some(uti) => {
                        match uti.as_str() {
                            "com.apple.coreaudio-format" => MediaType::Audio(uti),
                            _ => MediaType::Other(uti),
                        }
                    }
                    None => MediaType::Unknown,
                }
            }
```

2. Attachment path generation function may replace '\~' characters in original filenames
https://github.com/ReagentX/imessage-exporter/blob/c73bc4d666da3a543708b553c9106acf89454e81/imessage-database/src/tables/attachment.rs#L251-L257
Line 254 may replace not only the '\~' characters at the beginning of path strings, but also those in the original filenames at the end of it, resulting in errors when copying attachments
Suggestion: It might be better to ensure that only the first '\~' character is replaced:
```
            return format!("{}{}", &home(), &path[1..]);
```

3. "chat_recoverable_message_join" table did not exist until iOS 16(at least not for iOS 15.4), but iMessage reply feature is supported since iOS 14. 
https://github.com/ReagentX/imessage-exporter/blob/c73bc4d666da3a543708b553c9106acf89454e81/imessage-database/src/tables/table.rs#L79
https://github.com/ReagentX/imessage-exporter/blob/c73bc4d666da3a543708b553c9106acf89454e81/imessage-database/src/tables/messages.rs#L151-L164
This fallback SQL ignores the "message" table altogether, causing html outputs of iOS 14/15 backup not to show any reply signs at all. 
Suggestion: adding another fallback SQL statement to the function below(and also the "stream_rows" function under it) would be good:
```
        .unwrap_or(db.prepare(&format!(
            "SELECT
                 *,
                 c.chat_id,
                 (SELECT COUNT(*) FROM {MESSAGE_ATTACHMENT_JOIN} a WHERE m.ROWID = a.message_id) as num_attachments,
                 (SELECT NULL) as deleted_from,
                 (SELECT COUNT(*) FROM {MESSAGE} m2 WHERE m2.thread_originator_guid = m.guid) as num_replies
             FROM
                 message as m
                 LEFT JOIN {CHAT_MESSAGE_JOIN} as c ON m.ROWID = c.message_id
             ORDER BY
                 m.date;
            "
        ))
```

4. "href" attributes of anchors for attachments may not work if a relative path to the backup is specified on the command line, causing attachments failed to display or be downloaded in html pages
https://github.com/ReagentX/imessage-exporter/blob/c73bc4d666da3a543708b553c9106acf89454e81/imessage-exporter/src/exporters/html.rs#L439-L451
Suggestion: consider removing "file://" substring. Browsers can handle it well. 

5. A pure suggestion: It's very easy to get it running directly on jailbroken iOS devices. The logic is literally the same as on Mac's, just change the path to the database. I've made a fork and managed to get it working flawlessly on my iPhone:
[imessage-exporter for iOS](https://github.com/rllbe/imessage-exporter)
It would be nice if you make it officially supported!

**Comment by ReagentX** [comment]
Tracking in #155, #156, #157, and #158.
